typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1302 @@
+          Tast_iterator.default_iterator.expr self e);
+    }
+  in
+  f iter;
+  !ll
+
+(* The following function marks all open variables in a class/module as shared,
+   as well as returning a UF reflecting all those shared usage. *)
+let mark_shared_open_variables ienv f _loc =
+  let ll = open_variables ienv f in
+  let ufs =
+    List.map
+      (fun (paths, maybe_unique) ->
+        (* the following force is not needed, because when UA is done on the
+           module/class, maybe_paths_of_ident will force free variables to
+           shared, because ienv given to it will not include the outside
+           variables. We nevertheless force it here just to be sure *)
+        let shared =
+          Usage.SharedUnique
+            (SharedUnique.force_boundary maybe_unique FreeVariableOfModClass)
+        in
+        let ufs = List.map (fun path -> UF.singleton path shared) paths in
+        UF.seqs ufs)
+      ll
+  in
+  UF.seqs ufs
+
+(* There are two modes our algorithm will work at.
+
+   In the first mode, we care about if the expression can be considered as
+   alias, for example, we want `a.x.y` to return the alias of a.x.y in addition
+   to the usage of borrowing a and a.x. Note that a.x.y is not included in the
+   usage, and the caller is responsible to mark a.x.y if it is used.
+
+   In the second mode, we don't care about if the expression can be considered
+   as alias. Checking a.x.y will return the usage of borrowing a and a.x, and
+   using a.x.y. This mode is used in most occasions. *)
+
+(* the following function corresponds to the second mode *)
+let rec check_uniqueness_exp_ exp (ienv : Ienv.t) : UF.t =
+  let loc = exp.exp_loc in
+  match exp.exp_desc with
+  | Texp_ident _ -> (
+      match check_uniqueness_exp' exp ienv with
+      | Some (paths, modes), uf ->
+          let occ = { Occurrence.loc; reason = DirectUse } in
+          let maybe_unique = (modes, occ) in
+          UF.seq uf (mark_maybe_unique paths maybe_unique)
+      | None, uf -> uf)
+  | Texp_constant _ -> UF.empty
+  | Texp_let (_, vbs, exp') ->
+      let ienv', uf = check_uniqueness_value_bindings_ vbs ienv in
+      let uf' = check_uniqueness_exp_ exp' (Ienv.seq ienv ienv') in
+      UF.seq uf uf'
+  | Texp_function { param; cases } ->
+      (* `param` is only a hint not a binder;
+         actual binding done in cases by Tpat_var and Tpat_alias *)
+      let value =
+        MatchSingle ([ UF.Path.fresh_root_of_ident param ], loc, None)
+      in
+      let uf = check_uniqueness_cases value cases ienv in
+      (* we are constructing a closure here, and therefore any borrowing of free
+         variables in the closure is in fact using shared. *)
+      let uf' =
+        UF.map
+          (function
+            | BorrowedShared (MaybeShared _ as u) ->
+                (* only implicit borrowing lifted. *)
+                SharedUnique
+                  (SharedUnique.Shared (BorrowedShared.extract_occurrence u))
+            | _ -> Unused)
```

Comment:
The root node has the highest mode in the tree (think `Unused`), and `par Unused x` gives `x` - I can't see the issue here?

Reply:
Yeah but right after the `UF.map`, the result is `par`-ed with the original `uf`. So that seems fine?

Reply:
It might be correct but I don't think it's fine. The correctness is not at all obvious. I really think it would be better to have a simple `UF.map` function that only works on monotonic functions, then to change this function to be monotonic by returning it's input in the `_` case, and then getting rid of the call to `par` afterwards as it won't be needed.